rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+      }
+    }
+  ]
+}
+```
+
+Instead of the customer having to look up and put together a JSON blob like this and then verifying and then inserting, we can also give them
+the option to do this using functions that would generate this blob.
+
+```sql
+-- User creates policy
+select * from iasql_create_policy(
+    'some_policy_name',
+    '2012-10-17' -- nullable, defaults to 2012-10-17
+    );
+
+-- User creates statement and attaches to the policy
+select * from iasql_create_policy_statement(
+        '<policy_id>',
+        'allow',
+        'aws:arn:123456789012:bucket-resource',
+        's3:GetObjects', -- comma separated list of actions
+        '123456789012' -- comma separated list of principals
+    );
+
+-- User creates bucket with the policy
+insert into bucket (
+                name,
+                policy_document, -- nullable if policy_id is provided
+                policy_id -- nullable if policy_document is provided
+            ) values ('some_bucket', NULL, '<policy_id>'); -- triggers a function that would generate the policy document, given the policy_id and store in the policy_document field
+```
+
+Perhaps these functions could even be used together like so
+    
+```sql
+-- User creates bucket with the policy
+insert into bucket (
+                name,
+                policy_document, -- nullable if policy_id is provided
+                policy_id -- nullable if policy_document is provided
+            ) values (
+                      'some_bucket',
+                      NULL,
+                      iasql_create_policy( -- returns policy ID after creation
+                          'some_policy_name',
+                          '2012-10-17', -- nullable, defaults to 2012-10-17
+                          iasql_create_policy_statement -- returns policy statement ID after creation
+                            '<policy_id>',
+                            'allow',
+                            'aws:arn:123456789012:bucket-resource',
+                            's3:GetObjects', -- comma separated list of actions
+                            '123456789012' -- comma separated list of principals
+                            ), -- Statement 1
+                          iasql_create_policy_statement(...), -- Statement 2
+                          ...
+                          iasql_create_policy_statement(...) -- Statement N
+                    );
+```
+
+### ER diagram for s3 example
+![Proposed model for s3](./assets/004/s3_example.png)
+
+And for Updating, we can equally have something like this:
+
+```sql
+UPDATE bucket set policy_id = iasql_create_policy( -- returns policy ID after creation
+                              'some_policy_name',
+                              '2012-10-17', -- nullable, defaults to 2012-10-17
+                              iasql_create_policy_statement -- returns policy statement ID after creation
+                                '<policy_id>',
```

Comment:
How would this inner policy statement function get the policy ID if it is created before the policy record?

Reply:
It's an error there. <policy_id> would not be part of that insert query.

Reply:
Got it.